myokit/formats/diffsl/_exporter.py

```diff
@@ -0,0 +1,523 @@
+
+            # Store (initial) name for var
+            var_to_name[var] = name
+
+        # Check for conflicts with known keywords
+        from . import keywords
+
+        needs_renaming = collections.OrderedDict()
+        for keyword in keywords:
+            needs_renaming[keyword] = []
+
+        pace = model.binding('pace')  # Reserve 'Vc' for pace variable
+        if pace is not None:
+            needs_renaming['Vc'] = []
+
+        # Find naming conflicts, create inverse mapping
+        name_to_var = collections.OrderedDict()
+        for var, name in var_to_name.items():
+
+            # Known conflict?
+            if name in needs_renaming.keys():
+                needs_renaming[name].append(var)
+                continue
+
+            # Test for new conflicts
+            var2 = name_to_var.get(name, None)
+            if var2 is not None:
+                needs_renaming[name] = [var2, var]
+                continue
+
+            name_to_var[name] = var
+
+        # Resolve naming conflicts
+        for name, variables in needs_renaming.items():
+            # Add a number to the end of the name, increasing it until it's
+            # unique
+            i = 1
+            root = name
+            for var in variables:
+                name = f'{root}{uscore}{i}'
+                while name in name_to_var:
+                    i += 1
+                    name = f'{root}{uscore}{i}'
+                var_to_name[var] = name
+                name_to_var[name] = var
+
+        if pace is not None:
+            var_to_name[pace] = 'Vc'
+
+        return var_to_name
+
+    def _get_time_factor(self, time_var):
+        """
+        Get factor to convert time to ms.
+        time_units * time_factor means going from time_units to ms.
+        """
+        time_unit = time_var.unit()
+
+        if time_unit is None:
+            return None, None
+
+        if time_unit == myokit.units.ms:
+            return None, None
+
+        time_factor = time_factor_inv = None
+
+        try:
+            time_factor = myokit.Number(
+                myokit.Unit.conversion_factor(time_unit, myokit.units.ms)
+            )
+            time_factor_inv = myokit.Number(
+                1 / time_factor.eval(), 1 / time_factor.unit()
+            )
+        except myokit.IncompatibleUnitError:
+            warnings.warn(
+                'Unable to convert time units '
+                + str(time_unit)
+                + ' to recommended units of ms.'
+            )
+
+        return time_factor, time_factor_inv
+
+    def _guess_currents(self, model):
+        """
+        Tries to make a list of membrane currents. Removes potentials
+        from the guessed list.
+        """
+        unmatch_units = [
+            myokit.units.mV,
+            myokit.units.V,
+        ]
+        guessed_currents = guess.membrane_currents(model)
+        return [x for x in guessed_currents if x.unit() not in unmatch_units]
```

Comment:
How well is this doing?
How come there are potentials in the list?

Reply:
I put the extra checks in after trying to convert the [full2-voltage-clamp-ikr-linleak.mmt](https://github.com/CardiacModelling/VoltageClampModel/blob/master/mmt-model-files/full2-voltage-clamp-ikr-linleak.mmt) model. I think what's happening is that `guess.membrane_currents()` thinks the variables on the RHS of membrane potential are all currents, but that model has `dot(V) = (Vp - V + voffset_eff) / tau - Iion / cm`.

Reply:
Makes sense! So the `guess` method is going to be quite wrong here 😓 
What exactly do we want the outputs to be here?
Some options:
- Manually list all things you want as outputs when calling the exporter
- Keep guessing currents, but annotate `i_ion` with `label cellular_current` in your models to find currents that are part of i_ion instead of ones that are part of V
- List all intermediary variables as outputs (`model.variables(inter=True, deep=False)`)

Reply:
I think for the exporter here, where the interest is things to display in a web browser app or suchlike, then would be fine to manually list all the things you want as outputs.

Reply:
I'm happy for that!

To be honest, I was always imagining the DiffSL export as a partially manual thing, similar to e.g. matlab and C export where the end result is something that runs but will almost definitely be adapted before actual use.

If we're imagining a fully automated pipeline from myokit models (in some specific form?) to the web-app, then it might be a good idea to build this *outside* of Myokit, so as a program that uses Myokit (possibly inheriting from a more generic diffsl export provided _in_ myokit).

Worth having a quick chat about?

Reply:
At this stage I don't think we need anything automated, might be nice one day when we have a website where you can upload any CellML or something, but at present these are going to be fairly carefully manually selected things where we only have O(10) of them for the first version of artefact web app. I'd keep it relatively simple and at least within myokit it all stays in sync and supported automatically.

Reply:
I guess there's already some post-processing being done adding the inputs and the protocol into the exported file. We could leave the outputs empty (or fill it with everything) and let the post-processing decide what to add (or remove).

I think if we want to specify the outputs manually, we should probably also allow manually specifying the inputs and the protocol as well. They can default to empty if they are not given.

Reply:
For manual inputs, the interface could be something like:

```python

inputs = ['voltage_clamp.E_leak_est', 'voltage_clamp.Cm_est']
# inputs will be translated to: 
#    in = [voltageClampELeakEst, voltageClampCmEst]

outputs = ['voltage_clamp.I_out', 'membrane.V']
# outputs will b translated to: 
#    out_i { voltageClampIOut, membraneV }

protocol = [[-80, 40, -40], [500, 500, 500], [false, false, false]]
# protocol = [[voltages], [durations], [is_ramp]], which will be translated to:
#    enginePace { -80 + 120 * heaviside(t - 500) - 80 * heaviside(t - 1000) }

# And run as:
model = myokit.load_model('/path/to/model.mmt')
exporter = myokit.formats.easyml.DiffSLExporter()
exporter.model('/path/to/model.diffsl', model, protocol=protocol, inputs=inputs, outputs=outputs)
```

Reply:
Myokit's protocols don't support ramps (yet): #802

Reply:
For now, I'll leave the inputs and outputs empty, and add an example in the comments showing how to populate them (there's already an example for the protocol). The website currently does post-processing to add the inputs, protocol, and outputs. It's not clear if adding them to the generated model will be a useful feature as they are going to be modified after.